File: python/219.py
class Solution1(object):
    def containsNearbyDuplicate(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: bool
        """
        checkDict = {}
        for i in range(len(nums)):
            if nums[i] in checkDict:
                if i - checkDict[nums[i]] <= k:
                    return True
            checkDict[nums[i]] = i
        return False

class Solution2(object):
    def containsNearbyDuplicate(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: bool
        """
        if k <= 0:
            return False
        if k > len(nums):
            k = len(nums)
        checkSet = set(nums[:k:])
        if len(checkSet) < k:
            return True
        for i in range(k, len(nums)):
            if nums[i] in checkSet:
                return True
            else:
                checkSet.discard(nums[i - k])
                checkSet.add(nums[i])
        return False

File: java/219.java
public class Solution1 {
//hashmap
    public static boolean containsNearbyDuplicate(int[] nums, int k) {
        HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();
        for (int i = 0; i < nums.length; i++){
        	if (map.containsKey(nums[i])){
        		if (i - map.get(nums[i]) <= k){
        			return true;
        		}
        		else{
        			map.put(nums[i], i);
        		}
        	}
        	else{
        		map.put(nums[i], i);
        	}

        }
        return false;
    }
}

public class Solution2 {
//set
    public static boolean containsNearbyDuplicate(int[] nums, int k) {
        Set<Integer> set = new HashSet<Integer>();
        for (int i = 0; i < nums.length; i++){
        	if (i > k) set.remove(nums[i - k - 1]);
        	if (!set.add(nums[i])) return true;
        }
        return false;
    }
}
